### PR TITLE
Add regrade support for bundle item submissions

### PR DESCRIPTION
**Pull request by nathanchrs**
Implements #192.

Add bundle item submissions regrade endpoints for Uriel, and regrade buttons in Raphael.

image
image

### CODE REVIEW

File: judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/BaseItemSubmissionStore.java [hunk 1]
```diff
@@ -69,43 +72,72 @@ public ItemSubmission upsertSubmission(
...
+            submissionModel.verdict = Verdict.PENDING_REGRADE.name();
+            submissionModel.score = null;
+            submissionDao.persist(submissionModel);
+        }
+
+        return submissions;
+    }
+
+    @Override
+    public void saveRegradeResult(String submissionJid, Grading grading) {
```
Comment (fushar) [inline judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/BaseItemSubmissionStore.java]: `updateGrading`

**Review by fushar — CHANGES_REQUESTED** [review]

File: judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/ItemSubmissionRegradeProcessor.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        for (ItemSubmission submission : submissions) {
+            Optional<Item> item = problemClient.getItem(submission.getProblemJid(), submission.getItemJid());
+            if (item.isPresent()) {
+                Grading grading = itemSubmissionGraderRegistry
+                        .get(item.get().getType())
+                        .grade(item.get(), submission.getAnswer());
+
+                itemSubmissionStore.saveRegradeResult(submission.getJid(), grading);
+
+                if (submission.getGrading().isPresent()) {
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/ItemSubmissionRegradeProcessor.java]: OK

File: judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/bundle/ContestItemSubmissionService.java [hunk 3]
```diff
@@ -49,4 +50,20 @@ ContestantAnswerSummaryResponse getAnswerSummaryForContestant(
             @QueryParam("contestJid") String contestJid,
             @QueryParam("username") Optional<String> username,
             @QueryParam("problemAlias") String problemAlias);
+
+    @POST
+    @Path("/{submissionJid}/regrade")
+    @Consumes(APPLICATION_JSON)
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/bundle/ContestItemSubmissionService.java]: Annotation removed

File: judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/programming/ContestSubmissionService.java [hunk 4]
```diff
@@ -36,14 +37,14 @@ SubmissionWithSourceResponse getSubmissionWithSourceById(
 
     @POST
     @Path("/{submissionJid}/regrade")
-    @Produces(APPLICATION_JSON)
+    @Consumes(APPLICATION_JSON)
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/programming/ContestSubmissionService.java]: Annotation removed

File: judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/bundle/ContestItemSubmissionService.java [hunk 5]
```diff
@@ -49,4 +50,20 @@ ContestantAnswerSummaryResponse getAnswerSummaryForContestant(
...
+    void regradeSubmission(
+            @HeaderParam(AUTHORIZATION) AuthHeader authHeader,
+            @PathParam("submissionJid") String submissionJid);
+
+    @POST
+    @Path("/regrade")
+    @Consumes(APPLICATION_JSON)
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/bundle/ContestItemSubmissionService.java]: Annotation removed

File: judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/programming/ContestSubmissionService.java [hunk 6]
```diff
@@ -36,14 +37,14 @@ SubmissionWithSourceResponse getSubmissionWithSourceById(
...
     void regradeSubmission(
             @HeaderParam(AUTHORIZATION) AuthHeader authHeader,
             @PathParam("submissionJid") String submissionJid);
 
     @POST
     @Path("/regrade")
-    @Produces(APPLICATION_JSON)
+    @Consumes(APPLICATION_JSON)
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-backends/uriel/uriel-api/src/main/java/judgels/uriel/api/contest/submission/programming/ContestSubmissionService.java]: Annotation removed

File: judgels-frontends/raphael/src/routes/uriel/contests/single/submissions/Bundle/VerdictTag/VerdictTag.tsx [hunk 7]
```diff
@@ -19,14 +20,19 @@ const verdictDisplayCode: { [verdict: string]: string | React.ReactNode } = {
...
   [Verdict.PENDING_MANUAL_GRADING]: 'MANUAL',
   [Verdict.WRONG_ANSWER]: 'WA',
 };
 
 export const VerdictTag: React.FunctionComponent<VerdictTagProps> = ({ verdict }) => {
   const intent = verdictIntentMap[verdict] || Intent.NONE;
-  const displayCode =
-    verdict === Verdict.PENDING_MANUAL_GRADING ? <Icon icon="pause" /> : verdictDisplayCode[verdict] || verdict;
+  let displayCode = verdictDisplayCode[verdict] || verdict;
+  if (verdict === Verdict.PENDING_REGRADE) {
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-frontends/raphael/src/routes/uriel/contests/single/submissions/Bundle/VerdictTag/VerdictTag.tsx:31]: OK

File: judgels-frontends/raphael/src/routes/uriel/contests/single/submissions/Bundle/modules/contestSubmissionActions.ts [hunk 8]
```diff
@@ -36,4 +36,22 @@ export const contestSubmissionActions = {
       return contestSubmissionBundleAPI.getLatestSubmissionsByUserForProblemInContest(token, contestJid, problemAlias);
     };
   },
+
+  regradeSubmission: (submissionJid: string) => {
+    return async (dispatch, getState, { contestSubmissionBundleAPI, toastActions }) => {
+      const token = selectToken(getState());
+      await contestSubmissionBundleAPI.regradeSubmission(token, submissionJid);
+
+      toastActions.showSuccessToast('Regrade in progress.');
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-frontends/raphael/src/routes/uriel/contests/single/submissions/Bundle/modules/contestSubmissionActions.ts]: Message changed to 'Submission regraded.'

File: judgels-frontends/raphael/src/routes/uriel/contests/single/submissions/Bundle/VerdictTag/VerdictTag.tsx [hunk 9]
```diff
@@ -19,14 +20,19 @@ const verdictDisplayCode: { [verdict: string]: string | React.ReactNode } = {
...
+    displayCode = <Icon icon="refresh" />;
+  } else if (verdict === Verdict.PENDING_MANUAL_GRADING) {
+    displayCode = <Icon icon="follower" />;
```
[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-frontends/raphael/src/routes/uriel/contests/single/submissions/Bundle/VerdictTag/VerdictTag.tsx:34]: image
To differentiate between pending submissions (awaiting regrade) and submissions that require manual grading. The `follower` icon shows that a person needs to manually look at the submission.

[the remark this answers is not in the document]
Comment (nathanchrs) [reply judgels-frontends/raphael/src/routes/uriel/contests/single/submissions/Bundle/VerdictTag/VerdictTag.tsx:34]: The `refresh` icon has been changed to `time` icon

File: judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/BaseItemSubmissionStore.java (hunk 1 above)
Comment (nathanchrs) [reply judgels-backends/sandalphon/sandalphon-client/src/main/java/judgels/sandalphon/submission/bundle/BaseItemSubmissionStore.java]: Renamed

**Review by fushar — APPROVED** [review]

**Pull request merged**